File: 2015/go/day10/day10.go
package day10

import (
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
)

//Day10 Elves Look, Elves Say
func Day10(inputFile string) error {
	input, err := ioutil.ReadFile(inputFile)
	if err != nil {
		return err
	}

	fmt.Println("Day 10: Elves Look, Elves Say")
	part1, err := part1(string(input))
	if err != nil {
		return err
	}
	fmt.Printf("\tPart 1: %d\n", len(part1))

	part2, err := part2(string(part1))
	if err != nil {
		return err
	}
	fmt.Printf("\tPart 2: %d\n", part2)

	return nil
}

func part1(input string) (string, error) {
	b := strings.Builder{}
	for i := 0; i < 40; i++ {
		last := '\u0000'
		count := 0

		for _, c := range input {
			if c == last {
				count++
			} else {
				if count > 0 {
					b.WriteString(strconv.Itoa(count))
					b.WriteRune(last)
				}
				last = c
				count = 1
			}
		}
		b.WriteString(strconv.Itoa(count))
		b.WriteRune(last)

		input = b.String()
		b.Reset()
	}

	return input, nil
}

func part2(input string) (int, error) {
	b := strings.Builder{}
	for i := 40; i < 50; i++ {
		last := '\u0000'
		count := 0

		for _, c := range input {
			if c == last {
				count++
			} else {
				if count > 0 {
					b.WriteString(strconv.Itoa(count))
					b.WriteRune(last)
				}
				last = c
				count = 1
			}
		}
		b.WriteString(strconv.Itoa(count))
		b.WriteRune(last)

		input = b.String()
		b.Reset()
	}

	return len(input), nil
}

File: 2015/rust/src/day10.rs
use simple_error::require_with;
use std::error::Error;
use std::fs::File;
use std::io::prelude::*;
use std::time::Instant;

pub fn day10(input_file: &str) -> Result<(), Box<dyn Error>> {
    println!("Day 10: Elves Look, Elves Say");

    let mut f = File::open(input_file)?;
    let mut input = String::new();
    f.read_to_string(&mut input)?;

    let start = Instant::now();
    let (sol1, cont) = part1(&input)?;
    println!("\tPart 1: {} ({:?})", sol1, start.elapsed());
    let second = Instant::now();
    println!("\tPart 2: {} ({:?})", part2(&cont)?, second.elapsed());
    println!("\t\t Completed in {:?}", start.elapsed());
    Ok(())
}

fn part1(input: &str) -> Result<(i32, String), Box<dyn Error>> {
    let mut input = input.to_owned();
    for _ in 0..40 {
        let mut output = String::new();
        let mut iter = input.chars();
        let mut last = require_with!(iter.next(), &format!("Invalid input: {}", input));
        let mut count = 1;
        while let Some(num) = iter.next() {
            if num == last {
                count += 1;
            } else {
                output.push_str(&format!("{}{}", count, last));
                count = 1;
                last = num;
            }
        }
        output.push_str(&format!("{}{}", count, last));
        input = output;
    }
    return Ok((input.len() as i32, input));
}

fn part2(input: &str) -> Result<i32, Box<dyn Error>> {
    let mut input = input.to_owned();
    for _ in 40..50 {
        let mut output = String::new();
        let mut iter = input.chars();
        let mut last = require_with!(iter.next(), &format!("Invalid input: {}", input));
        let mut count = 1;
        while let Some(num) = iter.next() {
            if num == last {
                count += 1;
            } else {
                output.push_str(&format!("{}{}", count, last));
                count = 1;
                last = num;
            }
        }
        output.push_str(&format!("{}{}", count, last));
        input = output;
    }
    return Ok(input.len() as i32);
}
